src/lib.rs

```diff
@@ -202,17 +221,35 @@ impl<T: Send> ThreadLocal<T> {
     {
         let thread = thread_id::get();
         if let Some(val) = self.get_inner(thread) {
-            return Ok(val);
+            return Ok(val)
         }
 
         Ok(self.insert(create()?))
     }
 
+    /// Returns a mutable element for the current thread, or creates it if it
+    /// doesn't exist. If `create` fails, that error is returned and no
+    /// element is added.
+    pub fn get_mut_or_try<F, E>(&mut self, create: F) -> Result<&mut T, E>
```

Comment:
I'm not too happy with this implementation. In order to adhere to the goals of this PR (mutable methods that enforce borrow checking) I've used a shabby work around to ensure that self is not mutably borrowed twice. This means that it's technically doing up-to three `get_inner`s, which seems a bit dirty.

Reply:
No longer a concern since the latest commit.